Source/Core/Common/Debug/CodeTrace.cpp

```diff
@@ -0,0 +1,391 @@
+
+  // Convert sp, rtoc, and ps to r1, r2, and F#. ps is handled like a float operation.
+  static const std::regex replace_sp("(\\W)sp");
+  instr = std::regex_replace(instr, replace_sp, "$1r1");
+  static const std::regex replace_rtoc("rtoc");
+  instr = std::regex_replace(instr, replace_rtoc, "r2");
+  static const std::regex replace_ps("(\\W)p(\\d+)");
+  instr = std::regex_replace(instr, replace_ps, "$1f$2");
+
+  // Pull all register numbers out and store them. Limited to Reg0 if ps operation, as ps get
+  // too complicated to track easily.
+  // ex: add r4, r5, r6 -> r4 = Reg0, r5 = Reg1, r6 = Reg2. Reg0 is always the target register.
+  static const std::regex regis(
+      "\\W([rfp]\\d+)[^r^f]*(?:([rf]\\d+))?[^r^f\\d]*(?:([rf]\\d+))?[^r^f\\d]*(?:([rf]\\d+))?",
+      std::regex::optimize);
+
+  if (std::regex_search(instr, match, regis))
+  {
+    tmp_attributes.reg0 = match.str(1);
+    if (match[2].matched)
+      tmp_attributes.reg1 = match.str(2);
+    if (match[3].matched)
+      tmp_attributes.reg2 = match.str(3);
+    if (match[4].matched)
+      tmp_attributes.reg3 = match.str(4);
+
+    if (instruction.memory_target)
+    {
+      tmp_attributes.memory_target = instruction.memory_target;
+      tmp_attributes.memory_target_size = GetMemoryTargetSize(instr);
+
+      if (StringBeginsWith(instr, "st") || StringBeginsWith(instr, "psq_s"))
+        tmp_attributes.is_store = true;
+      else
+        tmp_attributes.is_load = true;
+    }
+  }
+
+  return tmp_attributes;
+}
+
+TraceOutput CodeTrace::SaveCurrentInstruction() const
+{
+  // Quickly save instruction and memory target for fast logging.
+  TraceOutput output;
+  const std::string instr = PowerPC::debug_interface.Disassemble(PC);
+  output.instruction = instr;
+  output.address = PC;
+
+  if (IsInstructionLoadStore(output.instruction))
+    output.memory_target = PowerPC::debug_interface.GetMemoryAddressFromInstruction(instr);
+
+  return output;
+}
+
+bool CompareMemoryTargetToTracked(const std::string& instr, const u32 mem_target,
+                                  const std::set<u32>& mem_tracked)
+{
+  // This function is hit often and should be optimized.
+  auto it_lower = std::lower_bound(mem_tracked.begin(), mem_tracked.end(), mem_target);
+
+  if (it_lower == mem_tracked.end())
+    return false;
+  else if (*it_lower == mem_target)
+    return true;
+
+  // If the base value doesn't hit, still need to check if longer values overlap.
+  const u32 range = GetMemoryTargetSize(instr) - 1;
+
+  return *it_lower <= mem_target + range;
+}
+
+AutoStepResults CodeTrace::AutoStepping(bool continue_previous, AutoStop stop_on)
+{
+  AutoStepResults results;
+
+  if (!CPU::IsStepping() || m_recording)
+    return results;
+
+  TraceOutput pc_instr = SaveCurrentInstruction();
+  const InstructionAttributes instr = GetInstructionAttributes(pc_instr);
+
+  // Not an instruction we should start autostepping from (ie branches).
+  if (instr.reg0.empty())
+    return results;
```

Comment:
I think this should be checking if it's continuing a previous run, otherwise it won't do anything if it stopped on a branch (timeout) or tracing from the register widget.
```C++
if (instr.reg0.empty() && !continue_previous)
```

Reply:
Good catch